### DISCUSSION TITLE
Printing Custom Text To LCD

### DISCUSSION THREAD

**Discussion by domiluci** [forntoh/LcdMenu#45] [Q&A] [answered]
Hey there, apologies for bugging you with this, but I'm having some trouble. I'm trying to print some custom, non-menu text to the LCD during `void setup()` (which will pave the way to adding custom text in later functions, callbacks, and such). But when I run the program, I can `init()` the LCD, and I get serial output in `void setup()`, but only up until line #X where I attempt to turn on the backlight and print a message to the LCD. It's like the program just stops right there.

The sad part is I'm sure this is one of those "oh... duh!" moments and I'm just doing it wrong. The thing is I had that "HELLO!" message working in that same spot in `setup()` when I was just tinkering around with `<LiquidCrystal_I2C.h>`, which I will comment out so you can see. But being able to print text besides the menus is sort of important to this project. Hopefully I just missed something and it's an easy fix, cause this library has been exactly what I needed to simplify this project.

Please forgive the messiness. It's a rough work in progress. Also, `LcdMenu.h` is in quotes because I'm using an IDE and simulator that doesn't have your library available. So I had to add the files to the project locally.

Here's the code:

```
#include <Wire.h>
//#include <LiquidCrystal_I2C.h>   // INACTIVE (For Reference)
#include "LcdMenu.h"

///// VARIABLES /////
// Declare LCD Variables
const uint8_t LCD_W = 20;  // Width in Characters
const uint8_t LCD_H = 4;     // Height in Rows
const uint8_t LCD_Address = 0x27;  // I2C Address

//LiquidCrystal_I2C LCD(LCD_Address, LCD_W, LCD_H); // INACTIVE (For Reference)

// Declare Menu Variables
// Declare the main menu
extern MenuItem MainMenu[];
extern MenuItem SettingsMenu[];

// Define the main menu
MenuItem MainMenu[] = {ItemHeader(),
                       MenuItem("Manual Mode"),
                       MenuItem("Preset Mode"),
                       MenuItem("Settings"),
                       ItemFooter()};

LcdMenu menu(LCD_H, LCD_W);

void setup() { // MAIN SETUP
  // Start Serial Output
  Serial.begin(115200);

  // Power On
  Serial.println("POWERING ON...");
    
  // Initialize The LCD
  //LCD.init();   // OLD WAY
  menu.lcd->init();   // NEW WAY

  // Serial Debugging
  Serial.print("LCD: "); Serial.print(LCD_W);
  Serial.print("X"); Serial.print(LCD_H);
  Serial.print(" at 0x"); Serial.println(String(LCD_Address, HEX));
  
  // Print initial message to the LCD.

  // OLD WAY USING THE STRAIGHT LQ_I2C LIBRARY (For Reference)
  //LCD.backlight();
  //LCD.setCursor(7, 1);
  //LCD.print("HELLO!");
  //Serial.println("HELLO!");
  //delay(2000);
  //LCD.clear();

  // NEW WAY USING YOUR LCDMENU LIBRARY
  // This is the place where the program stops and nothing shows on the LCD
  menu.toggleBacklight();
  menu.lcd->setCursor(7, 1);
  menu.lcd->print("HELLO!");
  Serial.println("HELLO!");
  delay(2000);

  // Print MainMenu to LCD
  menu.setupLcdWithMenu(LCD_Address, MainMenu);   // This part works if I remove the above code

}
```

So...what am I missing? What am I doing wrong? Thanks for the help!

**Comment by forntoh** [comment, marked as answer]
Please check the answer to this discussion #12

> Now there is a `hide` and a `show` function, calling hide will clear the display.

Also, note that you can't do anything with the menu without first setting it up (initializing) i.e. `menu.setupLcdWithMenu` should come before everything else.

**Reply by forntoh** [reply]
Find all the methods and fields [here](https://forntoh.gitbook.io/lcdmenu/reference/lcdmenu)

**Reply by domiluci** [reply]
Before I posted, I went through all the discussions, so I knew about the hide() and show() functions. I planned on using them later on.

A lot of why I asked was so I could figure out how to more directly manipulate the LCD through your LcdMenu library, as you already include the LiquidCrystal_I2C library. So I wanted to avoid declaring the same LCD twice just to use it in the .ino, mostly to prevent issues and/or interactions, especially as I have other I2C devices I'll be using. But also so I could add extra features that work in conjunction with your library.

See! I knew I was missing something simple, and I don't know why I didn't catch that earlier! But now that I'm going through your LcdMenu.h code, I can see where you declare the LCD in setupLcdWithMenu(). So I added the declaration code into my .ino, and now it works great. I'll edit this post and attach that code as an answer once I'm home in front of my code.

Also, thank you for the link with all the methods and fields. I didn't see that earlier, but it's extremely helpful!

While I have you here, I do have one more inquiry in the line of customization. Is there a way to skip rows of a menu, both for the menu item and the cursor? For instance, you have a 3 row menu, but the menu items would go on rows 1, 2, and 4, and the cursor would also skip line 3. Cause I could add a blank menu item, totally. But it'd look neater for my application if that was possible. If not, I can always write that feature in. Just thought I'd ask first ☺️.

Thank you very much for your time and help. I really appreciate it!

**Reply by forntoh** [reply]
I'm glad that I was able to help. As per my thinking, the functionality which you want to include is very not very niche, most users of this library wouldn't really use that functionality. You can fork the repo and release your own version with that functionality.
_Also, please explain to me more about it (why would you need to skip a menu/why would you need to have a blank menu item), maybe I don't understand you well_

**Comment by Stefano-69** [comment]
Hi Domiluci

Many thanks for your example; it was exactly what I wanted to do, but I didn't know how.

In your code you can even comment the line "include<wire.h>" because it is redundant. 

//#include <Wire.h>  // INACTIVE (For Reference)
//#include <LiquidCrystal_I2C.h>   // INACTIVE (For Reference)
#include "LcdMenu.h"

Regards,

Stefano
